src/main/java/org/zalando/nakadi/cache/SimpleCache.java

```diff
@@ -0,0 +1,99 @@
+package org.zalando.nakadi.cache;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+public class SimpleCache
+        <T, Versioned extends VersionedEntity<Version>, Version extends Comparable>
+        implements Cache<T> {
+
+    private final CacheDataProvider<Versioned, Version> dataProvider;
+    private final Function<Versioned, T> rawDataConverter;
+
+    private final Map<String, Versioned> versionedData = new HashMap<>();
+    private final Object versionedDataLock = new Object();
+
+    private final ConcurrentHashMap<String, T> activeData = new ConcurrentHashMap<>();
+    private final List<Consumer<String>> invalidationListeners = new CopyOnWriteArrayList<>();
+
+    public SimpleCache(
+            final CacheDataProvider<Versioned, Version> dataProvider,
+            final Function<Versioned, T> rawDataConverter) {
+        this.dataProvider = dataProvider;
+        this.rawDataConverter = rawDataConverter;
+    }
+
+    private T loadItemLocked(final String key) {
+        // this operation is relatively long, therefore we are making it concurrent
```

Comment:
I didn't understand. How is this concurrent? You mean that `computeIfAbsent` doesn't block other concurrent calls to different keys?

Reply:
Yes, exactly